include/linux/bpf.h

```diff
@@ -58,119 +65,172 @@ enum bpf_cmd_id
+    uint64_t insns;               ///< Array of instructions
+    uint64_t license;      ///< Optional pointer to a string specifying the license (currently ignored on Windows).
+    uint32_t log_level;    ///< Logging level (currently ignored on Windows).
+    uint32_t log_size;     ///< Size in bytes of the log buffer.
+    uint64_t log_buf;      ///< Pointer to a buffer in which log info can be written.
+    uint32_t kern_version; ///< Kernel version (currently ignored on Windows).
+    uint32_t prog_flags;   ///< Not supported.
+} sys_bpf_prog_load_attr_t;
+
+typedef struct
+{
+    uint32_t target_fd;               ///< eBPF target to attach/detach to/from.
+    uint32_t attach_bpf_fd;           ///< File descriptor of program to attach to.
+    enum bpf_attach_type attach_type; ///< Type of program to attach/detach to/from.
+    uint32_t attach_flags;            ///< Flags affecting the attach operation.
+} sys_bpf_prog_attach_attr_t;
+
+typedef struct
+{
+    uint32_t prog_fd;       ///< File descriptor of program to run.
+    uint32_t retval;        ///< On return, contains the return value of the program.
+    uint32_t data_size_in;  ///< Size in bytes of input data.
+    uint32_t data_size_out; ///< Size in bytes of output data.
+    uint64_t data_in;       ///< Pointer to input data.
+    uint64_t data_out;      ///< Pointer to output data.
+    uint32_t repeat;        ///< Number of times to repeat the program.
+    uint32_t duration;      ///< Duration in milliseconds to run the program.
+    uint32_t ctx_size_in;   ///< Size in bytes of input context.
+    uint32_t ctx_size_out;  ///< Size in bytes of output context.
+    uint64_t ctx_in;        ///< Pointer to input context.
+    uint64_t ctx_out;       ///< Pointer to output context.
+    uint32_t flags;         ///< Flags (currently 0).
+    uint32_t cpu;           ///< CPU to run the program on.
+    uint32_t batch_size;    ///< Number of times to run the program in a batch.
+    uint32_t _pad0;
+} sys_bpf_prog_run_attr_t;
+
+typedef struct
+{
+    uint64_t pathname; ///< Path name for pinning.
+    uint32_t bpf_fd;   ///< File descriptor referring to the program or map.
+    uint32_t flags;    ///< Not supported.
+} sys_bpf_obj_pin_attr_t;
+
 /// Attributes used by BPF_OBJ_GET_INFO_BY_FD.
 typedef struct
 {
     uint32_t bpf_fd; ///< File descriptor referring to an eBPF object.
-    uint64_t info;   ///< Pointer to memory in which to write the info obtained.
-
     /**
      * @brief On input, contains the maximum number of bytes to write into the info. On output, contains
      * the actual number of bytes written.
      */
     uint32_t info_len;
-} bpf_obj_info_attr_t;
+    uint64_t info; ///< Pointer to memory in which to write the info obtained.
+} sys_bpf_obj_info_attr_t;
 
 /// Attributes used by BPF_LINK_DETACH.
 typedef struct
 {
     uint32_t link_fd; ///< File descriptor of link to detach.
-} bpf_link_detach_attr_t;
+} sys_bpf_link_detach_attr_t;
```

Comment:
I don't like the idea of always breaking source compat.  I'd prefer to see some ifdef that allows source compat using the original names.

Reply:
This would be a significant amount of work on top of maintaining ABI compat and it's not clear to me who would benefit from it. If I was working on a C code base which wants to directly call `bpf()` with Linux source compat there is a very high likelihood that I'm already pulling in upstream `uapi/linux/bpf.h` anyways. It's certainly not required to get ebpf-go going, so I'd really rather not do this.

I'd even go so far to hide these types from users altogether. Only expose a forward declared `union bpf_attr` and the corresponding `bpf()` declaration and put them into a private header that is included from unit tests.

Reply:
Ok to defer until needed